### ISSUE TITLE
Differences in read counts from sorted and unsorted BAM files

### ISSUE THREAD

**Issue by s-andrews** [s-andrews/SeqMonk#99]
We've received the following report:

I am using SeqMonk 1.41 on Windows and samtools 1.7 on Ubuntu and am getting an inconsistent number of reads in SeqMonk after I sort a BAM file

It is from an Illumina single read fastq file aligned to the (bacterial) reference genome using Bowtie.

First, I sampled approximately 1M lines from the SAM file to speed up the tests.
```samtools view -h -s 0.07425 -S ORIGINAL01.sam -o SAMPLE01.sam```

Then I converted SAM to BAM
```$ samtools view -bS SAMPLE01.sam -o SAMPLE01.bam```

I then sorted the BAM file:
```$ samtools sort SAMPLE01.bam -o SORTED01.bam```

samtools shows that the number of reads and also the 
distribution of the alignment quality scores (which are all 0 or 255)
remains the same in the unsorted and sorted files:

```$ samtools view -cS SAMPLE01.sam
1001990
$ samtools view -c SAMPLE01.bam
1001990
$ samtools view -c SORTED01.bam
1001990

$ samtools view -S SAMPLE01.sam | cut -f5 | sort | uniq -c | sort -n -r
885924 255
116066 0
$ samtools view SAMPLE01.bam | cut -f5 | sort | uniq -c | sort -n -r
885924 255
116066 0
$ samtools view SORTED01.bam | cut -f5 | sort | uniq -c | sort -n -r
885924 255
116066 0
```
However, when I load the files into SeqMonk (using the default import parameters), I get the following read counts:

```
SAMPLE01.sam 885,906
SAMPLE01.bam 885,906  
SORTED01.bam 661,580
```

I am trying to figure out why the SeqMonk read count gets reduced in the sorted BAM file.  Note that for the option "Remove Duplicate Reads", I am selecting "No".

However, when I change that to "Yes, based on Start", the read counts become the same:

```
SAMPLE01.sam 616,392
SAMPLE01.bam 616,392
SORTED01.bam 616,392
```

I think this implies that in the sorted BAM file, some duplicates are being dropped when importing into SeqMonk even if Remove Duplicate Reads is set to "No".

I have seen other cases (where there are 13M reads) where the read count in the sorted file is reduced even more dramatically.  For example, in the original file (before I sampled 1M), I get 11.9M total reads from the unsorted BAM file, but 4.6M total reads from the sorted BAM file, and then when I do remove duplicates, the read counts for both unsorted and sorted drop to 3.14M.

My apologies for the length of this message.  I am not sure if this is due to a bug or due to my mistake or misunderstanding.  I am inexperienced as this is my first next generation sequencing project.

**Comment by s-andrews** [comment]
Obviously this isn't expected behaviour.  I've asked the original reporter for the data they used to see if we can specifically see what's happening.

With the default import parameters the only thing which should reduce the number of sequences imported would be the MAPQ filter, or the primary sequence filter.  Other than that everything should come in.

I'll try to reproduce this on a test data set here and see if we can also see this.

**Comment by s-andrews** [comment]
I did a quick test, and I can confirm something anomalous is happening:
[andrewss@rocks1 SortingBug]$ samtools view -h -s 0.01 -b -S ENCFF949JDD_TBP1_Rep1_WBcel235_bowtie2.bam -o unsorted.bam

```
[andrewss@rocks1 SortingBug]$ ls -ltr
total 1317376
-rw-r--r-- 1 andrewss bioinf 998818347 Jun  1 08:36 ENCFF949JDD_TBP1_Rep1_WBcel235_bowtie2.bam
-rw-rw-r-- 1 andrewss bioinf  10581977 Jun  1 08:38 unsorted.bam
[andrewss@rocks1 SortingBug]$ samtools view unsorted.bam | wc -l
228396
[andrewss@rocks1 SortingBug]$ samtools sort unsorted.bam -o sorted.bam
[andrewss@rocks1 SortingBug]$ samtools view sorted.bam | wc -l
228396
```

After importing the two using default parameters I then used the visible store importer to deduplicate them and then doing a summary report I get:

```
DataStore		Total Read Count

sorted.bam		190523
unsorted.bam		190791
sorted.bam_dedup	189089
unsorted.bam_dedup	189089
```

**Comment by s-andrews** [comment]
This does seem to be related to duplication somehow.  Most of the missing reads are present already and are being undercounted, but it's not all duplicates - some still remain.

I'll have to make a really small dataset and extract those reads out of the file to see what's going on.

**Comment by s-andrews** [comment]
This isn't a simple mis-import.  Something more complicated is happening.  If I take a simple example of a position which gives 2 reads in the unsorted file, and only 1 in the sorted, I can then extract out just the lines for that read from the sorted file (there are two), and if I import that - then I get the two expected reads.

The data in the file is therefore OK, and the parser, at least initially, is parsing it correctly, but something else is going on, either at the sorting, or collation stage which is causing a count to get lost.

**Comment by s-andrews** [comment]
I've added debugging to the ChromosomeDataStore which handles the initial collation.  The difference between the two imports is that in the unsorted we get two separate entries in the count and read vectors since the reads are not adjacent in the file.  For the sorted we get a single entry with a count of 2.  The count is definitely incremented correctly though, so it must be a later stage where the differnce happens.

**Comment by s-andrews** [comment]
Found it!  It's a bug in the collation of unaligned reads.  It's also a bit of a niche case, which is why it didn't get picked up by our tests.

The error was that when we were collating the positions and counts for unsorted data, instead of adding the existing count to the running total for each read, we were adding a fixed value of 1.  For unsorted data this would be correct nearly all of the time, since the chances of an adjacent identical read are so small (our test case turns out not to have any).

For sorted data we do a test to see if we need to collate.  If the data is perfectly sorted then we skip the collation and the bug is never triggerd.  The test file I used though has some instances where reads at identical positions are separated by a read with a different position, so you end up with multiple entries for multi-counts in the collated data.  This then collapses down when we do the collation and almost deduplicates them.

I've added a patch to fix this and will put out a new release to make this available and warn people who might have imported sorted BAM files that this might have affected them.

Thanks for spotting this.
